packages/utilities/src/asAsync.tsx

```diff
@@ -0,0 +1,147 @@
+
+import * as React from 'react';
+
+interface IAsAsyncState<TProps> {
+  component?: React.ComponentType<TProps> | undefined;
+  error?: Error;
+  placeholder?: JSX.Element;
+  onLoad?: () => void;
+  onError?: (error: Error) => void;
+}
+
+/**
+ * Represents the surface of a module which has a default export.
+ */
+export interface IModuleWithDefault<T> {
```

Comment:
1. We need some tests.
2. The design probably should be more like:

It would be nice to add a test so nobody breaks it. It may also be nice to have some documentation on usage.

```
const MyAsyncComponent = asAsync<IMyComponentProps>({
  load: () => import('MyComponent'),
  name: 'MyComponent', // defaults to 'default',
  placeholder: () => <Spinner />
});
```
The props you pass into `MyAsyncComponent` should be passthrough: 

```tsx
<MyAsyncComponent foo='bar' />
```

I think this simplifies usage a lot.

Reply:
https://github.com/jamiebuilds/react-loadable for reference :)

Reply:
Still confused. `placeholder` would be optional. In addition I could even imagine `placeholderDelay` to avoid rendering a placeholder until some set amount of time. I don't quite understand what you mean by "you can't keep them updated in your component's `render` method."

I simply suggested:

1. Take an options prop bag, rather than parameters. This will let you add features later, like "placeholderDelay" without reworking the param order.

2. The props that the resulting component should take are the same props that component should take. Adding a mapping function is not needed and just adds complexity where it isn't needed.

Reply:
ping

Reply:
I strongly disagree that the loading component should be baked into the decorated component; the parent should supply whatever loading component is appropriate for the given component. The loading component might also require completely divergent props from the async component, so I don't understand why we'd want it to take the same props.
Any sort of `placeholderDelay` can also just be handled by the chosed `placeholder` implementation, so that doesn't need to be part of this implementation.
`mapAsyncProps` exists to keep the call site as simple as possible; most consumers just want to pass their loaded function and be done.

Reply:
I wasn't suggesting to bake in a loading component. :) I was suggesting to allow it to be provided, like react-loadable.

But, as Ken pointed out, we should just use react-loadable rather than inventing another one.